pandas/core/reshape/concat.py

```diff
@@ -518,8 +518,11 @@ def _sanitize_mixed_ndim(
                         # to have unique names
                         name = current_column
                         current_column += 1
-
-                obj = sample._constructor({name: obj}, copy=False)
+                    obj = sample._constructor(obj, copy=False)
+                    if isinstance(obj, ABCDataFrame):
+                        obj.columns = range(name, name + 1, 1)
```

Comment:
What type of Index gets returned by the constructor? Shouldn't we be creating the RangeIndex at a lower level?

Reply:
`._constructor` with a `dict` would have returned an `Index` column with a single element `name`. With `_constructor`, there is currently no way to dictate the return type of the columns (if `_constructor` supports columns)